diff --git a/fs/proc/task_mmu.c b/fs/proc/task_mmu.c
index 7ca7ae383..e43c79249 100644
--- a/fs/proc/task_mmu.c
+++ b/fs/proc/task_mmu.c
@@ -899,20 +899,6 @@ static void __show_smap(struct seq_file *m, const struct mem_size_stats *mss)
 		   mss->anonymous_thp >> 10,
 		   mss->shmem_thp >> 10,
 		   mss->shared_hugetlb >> 10,
+#ifdef CONFIG_KSU_SUSFS_SUS_MAP
		if (vma->vm_file &&
-			unlikely(file_inode(vma->vm_file)->i_mapping->flags & BIT_SUS_MAPS) &&
-			susfs_is_current_proc_umounted())
-		{
-			seq_printf(m,
-				"Size:           %8lu kB\n"
-				"KernelPageSize: %8lu kB\n"
-				"MMUPageSize:    %8lu kB\n",
-				(vma->vm_end - vma->vm_start) >> 10,
-				4, 4);
-			goto bypass_orig_flow;
-		}
-#endif
 		   mss->private_hugetlb >> 10,
 		   mss->swap >> 10,
 		   (unsigned long)(mss->swap_pss >> (10 + PSS_SHIFT)),
@@ -881,7 +881,20 @@ static int show_smap(struct seq_file *m, void *v)
 		seq_print_vma_name(m, vma);
 		seq_putc(m, '\n');
 	}
-
+#ifdef CONFIG_KSU_SUSFS_SUS_MAP
+		if (vma->vm_file &&
+			unlikely(file_inode(vma->vm_file)->i_mapping->flags & BIT_SUS_MAPS) &&
+			susfs_is_current_proc_umounted())
+		{
+			seq_printf(m,
+				"Size:           %8lu kB\n"
+				"KernelPageSize: %8lu kB\n"
+				"MMUPageSize:    %8lu kB\n",
+				(vma->vm_end - vma->vm_start) >> 10,
+				4, 4);
+			goto bypass_orig_flow;
+		}
+#endif
 	seq_printf(m,
 		   "Size:           %8lu kB\n"
 		   "KernelPageSize: %8lu kB\n"
@@ -890,11 +903,26 @@ static int show_smap(struct seq_file *m, void *v)
 		   vma_kernel_pagesize(vma) >> 10,
 		   vma_mmu_pagesize(vma) >> 10);
 
+#ifdef CONFIG_KSU_SUSFS_SUS_MAP
+bypass_orig_flow:
+#endif
 	__show_smap(m, &mss);
 
+#ifdef CONFIG_KSU_SUSFS_SUS_MAP
+		if (vma->vm_file &&
+			unlikely(file_inode(vma->vm_file)->i_mapping->flags & BIT_SUS_MAPS) &&
+			susfs_is_current_proc_umounted())
+		{
+			seq_puts(m, "VmFlags: mr mw me");
+			seq_putc(m, '\n');
+			goto bypass_orig_flow2;
+		}
+#endif
 	arch_show_smap(m, vma);
 	show_smap_vma_flags(m, vma);
-
+#ifdef CONFIG_KSU_SUSFS_SUS_MAP
+bypass_orig_flow2:
+#endif
 	m_cache_vma(m, vma);
 
 	return 0;
